src/ufuncs.c

```diff
@@ -1339,6 +1339,67 @@ static void snap_func(char** args, npy_intp* dimensions, npy_intp* steps, void*
 }
 static PyUFuncGenericFunction snap_funcs[1] = {&snap_func};
 
+static char clip_by_rect_dtypes[6] = {NPY_OBJECT, NPY_DOUBLE, NPY_DOUBLE,
+                                      NPY_DOUBLE, NPY_DOUBLE, NPY_OBJECT};
+static void clip_by_rect_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                              void* data) {
+  char *ip1 = args[0], *ip2 = args[1], *ip3 = args[2], *ip4 = args[3], *ip5 = args[4];
+  npy_intp is1 = steps[0], is2 = steps[1], is3 = steps[2], is4 = steps[3], is5 = steps[4];
+  npy_intp n = dimensions[0];
+  npy_intp i;
+  GEOSGeometry** geom_arr;
+  GEOSGeometry* in1 = NULL;
+
+  CHECK_NO_INPLACE_OUTPUT(5);
+
+  if ((is2 != 0) | (is3 != 0) | (is4 != 0) | (is5 != 0)) {
+    PyErr_Format(PyExc_ValueError,
+                 "clip_by_rect function called with non-scalar parameters");
+    return;
+  }
+
+  double xmin = *(double*)ip2;
+  double ymin = *(double*)ip3;
+  double xmax = *(double*)ip4;
+  double ymax = *(double*)ip5;
+
+  // allocate a temporary array to store output GEOSGeometry objects
+  geom_arr = malloc(sizeof(void*) * n);
+  CHECK_ALLOC(geom_arr);
+
+  GEOS_INIT_THREADS;
+
+  for (i = 0; i < n; i++, ip1 += is1) {
+    /* get the geometry: return on error */
+    if (!get_geom(*(GeometryObject**)ip1, &in1)) {
+      errstate = PGERR_NOT_A_GEOMETRY;
+      destroy_geom_arr(ctx, geom_arr, i - 1);
+      break;
+    }
+
+    if (in1 == NULL) {
+      // in case of a missing value: return NULL (None)
+      geom_arr[i] = NULL;
+    } else {
```

Comment:
It might be good to guard for empty geometries here.  It looks like right now they tend to be coalesced to empty geometry collections by this and by `intersection`, but I wonder if we should just return a clone of the input so that we aren't changing geometry type or dimensions by this operation.

Reply:
See also the answer above. I certainly understand the concern, and it was also my first reaction (why can't the empty geometry not preserve its geometry type?). But we also don't do such check for other overlay operations like intersection etc, so this might warrant a more general discussion/change?